### PR TITLE
spatial/r3: add Triangle

### PR DESCRIPTION
**Pull request by soypat**
Please take a look.

Resolves part of #1791

### CODE REVIEW

File: spatial/r3/triangle.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+// Triangle represents a triangle in 3D space and
+// is composed by 3 vectors corresponding to the position
+// of each of the vertices. Ordering of these vertices
+// decides the "normal" direction.
+type Triangle struct {
+	// Triangle vertices. Ordering of these vertices decides Normal vector direction.
+	// Inverting ordering of two vertices inverts the resulting direction.
+	V [3]Vec
+}
```
Comment (sbinet) [inline spatial/r3/triangle.go:17]: just a drive-by comment:
any reason why we don't define it as:
```go
type Triangle [3]Vec
```

or, (better for the SSA pass):
```go
type Triangle struct {
    A, B, C Vec
}
```


do we expect to sub-slice `t.V`? or iterate over the vertices of a triangle?

File: spatial/r3/triangle.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+// Normal returns the vector with direction
+// perpendicular to the Triangle's face. The ordering
+// of the triangle vertices decides the normal's resulting
+// direction. Normal is not guaranteed to return a unit-norm vector.
+func (t Triangle) Normal() Vec {
+	s1, s2, _ := t.sides()
```
Comment (sbinet) [inline spatial/r3/triangle.go:24]: it seems (moderately) wasteful to ask for the computation of `Sub(t.V[0], t.V[2])` to drop it right there.
but ok.

**Review by sbinet — APPROVED** [review]
just a little stream of consciousness.

LGTM, perhaps add a little example?

File: spatial/r3/triangle.go (hunk 1 above)
Comment (soypat) [reply spatial/r3/triangle.go:17]: It's common to iterate over triangle vertices to make code more succinct. I can change it to `type Triangle [3]Vec`, which I prefer over the current implementation.

File: spatial/r3/triangle.go (hunk 2 above)
Comment (soypat) [reply spatial/r3/triangle.go:24]: Would the SSA not optimize the calculation? Maybe not today but it does seem like something that could done given it looks like a short enough path

**Comment by soypat** [comment]
@sbinet I've added an example for which I do not have the tools to validate. I have also added the `Centroid` method here and in the `r2` implementation.

**Review by sbinet — DISMISSED** [review]

**Comment by sbinet** [comment]
although, it'd be great if you could rebase off #1801 (to get the CI running)

**Comment by soypat** [comment]
@kortschak  I have added the tests requested with a handful of comments to help out readers. 

I managed to fix the icosphere example. Below are 0 and 7 subdivisions of an icosahedron.
image
image

File: spatial/r3/triangle.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+	return Cross(s1, s2)
+}
+
+// Degenerate returns true if all of triangle's vertices are
+// within tol distance of its longest side.
+func (t Triangle) Degenerate(tol float64) bool {
```
Comment (kortschak) [inline spatial/r3/triangle.go:27]: Consider renaming this to `IsDegenerate`.

File: spatial/r3/triangle.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+	ln := line{t[longIdx], t[(longIdx+1)%3]}
+	dist := ln.distance(t[(longIdx+2)%3])
+	return dist <= tol
+}
+
+// Area returns the surface area of the triangle.
+func (t Triangle) Area() float64 {
+	// Formula taken from Section 2, page 3.
+	// William M. Kahan (24 March 2000). "Miscalculating Area and Angles of a Needle-like Triangle"
+	// https://people.eecs.berkeley.edu/~wkahan/Triangle.pdf
```
Comment (kortschak) [inline spatial/r3/triangle.go:39]: Note that this is Heron's formula. I would probably link to the first (as well as this reference) given the ancient history of it. https://en.wikipedia.org/wiki/Heron%27s_formula.

```
// Heron's Formula, see https://en.wikipedia.org/wiki/Heron%27s_formula.
// Also see William M. Kahan (24 March 2000). "Miscalculating Area and Angles of a Needle-like Triangle"
// for more discussion. https://people.eecs.berkeley.edu/~wkahan/Triangle.pdf.
```

File: spatial/r3/triangle.go [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+		l1, l2 = l2, l1
+	}
+	if l3 < l2 {
+		l2, l3 = l3, l2
+		if l2 < l1 {
+			l1, l2 = l2, l1
+		}
+	}
+	return l1, l2, l3
+}
```
Comment (kortschak) [inline spatial/r3/triangle.go:95]: Used only once for `Area`, so place it immediately after that.

File: spatial/r3/triangle.go [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+	longIdx := 0
+	if len2[1] > longLen {
+		longLen = len2[1]
+		longIdx = 1
+	}
+	if len2[2] > longLen {
+		longIdx = 2
+	}
+	return longIdx
+}
```
Comment (kortschak) [inline spatial/r3/triangle.go:75]: Since this is only used once, for `Degenerate`, I'd place it immediately after that.

**Review by kortschak — COMMENTED** [review]
This is looking good.

**Comment by kortschak** [comment]
Can you rebase onto the current master to get the new linter.

File: spatial/r3/icosahedron_example_test.go [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+	var faces []r3.Triangle
+	for _, t := range triangles {
+		var face r3.Triangle
+		for i := 0; i < 3; i++ {
+			face[i] = vertices[t[i]]
+		}
+		faces = append(faces, face)
+	}
+	fmt.Println(faces)
+	// The 3D rendering of the icosphere is left as an exercise to the reader.
```
[the remark this answers is not in the document]
Comment (soypat) [reply spatial/r3/icosahedron_example_test.go:77]: Left a small pointer-comment, good enough?

File: spatial/r3/icosahedron_example_test.go [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+)
+
+// edgeIdx represents an edge of the icosahedron
+type edgeIdx [2]int
+
+// subdivideEdge takes the vertices list and indices first and second to
+// the vertices defining the edge that will be subdivided. lookup is a
+// table of all newly generated vertices from previous calls to subdivideEdge
+// so as to not duplicate vertices.
+func subdivideEdge(lookup map[edgeIdx]int, vertices []r3.Vec, first, second int) (int, []r3.Vec) {
```
[the remark this answers is not in the document]
Comment (soypat) [reply spatial/r3/icosahedron_example_test.go]: I reversed the order of the functions but left them before the example following suit of the euler angle example. Should I add them after the example or is this OK?

File: spatial/r3/icosahedron_example_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (kortschak) [reply spatial/r3/icosahedron_example_test.go:77]: Oh, sorry, no. OOB. This comment for the Barnes-Hut example was how the galaxy example for gio was instigated. This doesn't need to be here.

**Comment by soypat** [comment]
I rebased onto current master, which is [ci: install ragel to check rdf codegen](https://github.com/gonum/gonum/commit/84fd2bc2bbc4a322613c3e2e81647e3a9682bcc8)

**Comment by kortschak** [comment]
You're right, sorry. The linter change hasn't merged yet.

File: spatial/r3/icosahedron_example_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (kortschak) [reply spatial/r3/icosahedron_example_test.go]: The Euler example is a special case; it was in that order because people know (or think they know) what Euler Angles are and want to use them, and it was written as a warning. This example has helpers that are not (AFAICS) something to be copy-pasted into their code.

**Comment by soypat** [comment]
Refactored icosphere example:
- All triangle normals now point owtward from icosphere center
- Bug caused duplicate loose subdiving vertices has been fixed, also clearer return value
- Example has been moved to front of file.

**Comment by kortschak** [comment]
It might be worth inspecting the example's mesh for defect. You can do this by exporting it as an STL and inspecting it with meshmixer. If you don't have code for STL export, you can use this https://github.com/kortschak/stl.

**Review by kortschak — DISMISSED** [review]

**Comment by soypat** [comment]
Do not merge. I believe I found a bug with Area function.

**Comment by soypat** [comment]
Nevermind, all is good. I've added more tests to `TestTriangleArea` and a new `TestTriangleOrderedLengths` test.

**Comment by soypat** [comment]
Hey there peeps! I'm gonna go ahead and share something I'm really proud of: 
image
A sphere .stl file (**left**) was taken aas input and a signed distance function generated from it (**right**) using the [`Transform` type](https://github.com/gonum/gonum/pull/1802) and a generic rewrite of `kdtree` (just for kicks). I'm looking forward to exposing this functionality with gonum's upcoming `r3.Triangle` as an argument! This would allow users of `sdf` to edit .stl files as if they were native `sdf.SDF3` objects

File: spatial/r3/icosahedron_example_test.go [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+		result = append(result, newTriangles...)
+	}
+	return vertices, result
+}
+
+// subdivideEdge takes the vertices list and indices first and second which
+// refer to the edge that will be subdivided.
+// lookup is a table of all newly generated vertices from
+// previous calls to subdivideEdge so as to not duplicate vertices.
+func subdivideEdge(lookup map[edgeIdx]int, vertices []r3.Vec, first, second int) (int, []r3.Vec) {
```
Comment (sbinet) [inline spatial/r3/icosahedron_example_test.go:65]: perhaps also return the lookup table, to future proof any behaviour change in the guts of the `map` implementation?

Comment (kortschak) [reply spatial/r3/icosahedron_example_test.go:65]: I don't think that's necessary; maps can't change their reference value behaviour because of the compatibility promise.

**Review by kortschak — APPROVED** [review]
